File: web/src/lib/extractoCliente.ts
/**
 * Algoritmo de extracto — réplica de `client_report.py` / `func extrac.txt`.
 * `generarFilasExtracto` ≡ `generar_dataframe_extracto`
 * `calcularResumenExtracto` ≡ `calcular_resumen_extracto`
 */

/** Días de crédito por defecto; la deuda deja de generarse al cumplirse. */
export const DIAS_CREDITO_DEFAULT = 365;

export function parseDiasCredito(fechaFinal?: string | null): number {
  const raw = fechaFinal?.trim();
  if (raw && /^\d+$/.test(raw)) {
    const n = parseInt(raw, 10);
    return n > 0 ? n : DIAS_CREDITO_DEFAULT;
  }
  return DIAS_CREDITO_DEFAULT;
}

function addDays(d: Date, days: number): Date {
  const x = new Date(d);
  x.setDate(x.getDate() + days);
  return x;
}

export type RegistroExtracto = {
  fecha: Date;
  valor: number;
  tipo: string;
  referencia: string;
};

export type MetricasExtracto = {
  cuotas_generadas: number;
  cuotas_completas: number;
  cuotas_pagadas: number;
  cuotas_pendientes: number;
  total_pagado: number;
  deuda_total: number;
  ultimo_pago: string;
  dias_mora: number;
  cumplimiento_pct: number;
};

type RegistroModificado = {
  fecha: Date;
  valor: number;
  tipo: string;
  referencia: string;
};

export type FilaExtracto = {
  fechaProgramada: Date;
  fechaPago: Date | "";
  valorPagado: number;
  tipo: string;
  referencia: string;
};

function startOfDay(d: Date): Date {
  const x = new Date(d);
  x.setHours(0, 0, 0, 0);
  return x;
}

function daysBetween(a: Date, b: Date): number {
  const ms = startOfDay(b).getTime() - startOfDay(a).getTime();
  return Math.floor(ms / 86400000);
}

function dateRange(start: Date, end: Date): Date[] {
  const dates: Date[] = [];
  const cur = startOfDay(start);
  const endD = startOfDay(end);
  while (cur.getTime() <= endD.getTime()) {
    dates.push(new Date(cur));
    cur.setDate(cur.getDate() + 1);
  }
  return dates;
}

function formatFecha(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

function setCongelados(fechas: Iterable<string> | undefined): Set<string> {
  const out = new Set<string>();
  if (!fechas) return out;
  for (const raw of fechas) {
    const ymd = String(raw ?? "").slice(0, 10);
    if (/^\d{4}-\d{2}-\d{2}$/.test(ymd)) out.add(ymd);
  }
  return out;
}

function parseYmdLocal(ymd: string): Date {
  const [y, m, d] = ymd.split("-").map(Number);
  return new Date(y, m - 1, d);
}

function dateRangeSinCongelados(
  start: Date,
  end: Date,
  freeze: Set<string>,
): Date[] {
  return dateRange(start, end).filter((d) => !freeze.has(formatFecha(d)));
}

/** Días de mora calendario menos congelados posteriores al último pago. */
function diasMoraSinCongelados(
  ultimoPagoYmd: string,
  fin: Date,
  freeze: Set<string>,
): number {
  const inicioPago = parseYmdLocal(ultimoPagoYmd);
  const bruto = daysBetween(inicioPago, fin);
  if (bruto <= 0) return 0;
  const hasta = startOfDay(fin).getTime();
  const desde = inicioPago.getTime();
  let restar = 0;
  for (const ymd of freeze) {
    const t = parseYmdLocal(ymd).getTime();
    if (t > desde && t <= hasta) restar += 1;
  }
  return Math.max(0, bruto - restar);
}

/** Equivalente a `generar_dataframe_extracto` en client_report.py */
export function generarFilasExtracto(
  fechaInicio: Date,
  valorCuota: number,
  registros: RegistroExtracto[],
  diasCredito: number = DIAS_CREDITO_DEFAULT,
  fechaReferencia?: Date,
  fechasCongeladas: Iterable<string> = [],
): FilaExtracto[] {
  if (valorCuota <= 0) {
    throw new Error("valor_cuota inválido");
  }

  const freeze = setCongelados(fechasCongeladas);
  const inicio = startOfDay(fechaInicio);
  const fechaFinCredito = addDays(inicio, diasCredito - 1);
  let fin = startOfDay(fechaReferencia ?? new Date());
  if (fin.getTime() > fechaFinCredito.getTime()) {
    fin = fechaFinCredito;
  }

  const registrosModificados: RegistroModificado[] = registros
    .filter((r) => r.valor != null && !Number.isNaN(Number(r.valor)))
    .map((r) => ({
      fecha: startOfDay(r.fecha),
      valor: Number(r.valor),
      tipo: r.tipo || "",
      referencia: r.referencia || "",
    }));

  const total = registrosModificados.reduce((s, r) => s + r.valor, 0);
  const cuotasPagadasCeil = total > 0 ? Math.ceil(total / valorCuota) : 0;

  const diasRango = dateRangeSinCongelados(inicio, fin, freeze).length;
  if (cuotasPagadasCeil > diasRango) {
    const finExtendido = addDays(fin, cuotasPagadasCeil - diasRango);
    fin =
      finExtendido.getTime() > fechaFinCredito.getTime()
        ? fechaFinCredito
        : finExtendido;
  }

  const filas: FilaExtracto[] = dateRangeSinCongelados(inicio, fin, freeze).map(
    (d) => ({
      fechaProgramada: d,
      fechaPago: "",
      valorPagado: 0,
      tipo: "",
      referencia: "",
    }),
  );

  let saldo = 0;
  let pagosIdx = 0;

  for (let i = 0; i < filas.length; i++) {
    while (pagosIdx < registrosModificados.length && saldo < valorCuota) {
      const reg = registrosModificados[pagosIdx];
      let valor = reg.valor;

      while (valor + saldo >= valorCuota) {
        const faltaParaCuota = valorCuota - saldo;
        filas[i].valorPagado += faltaParaCuota;
        filas[i].fechaPago = reg.fecha;
        if (filas[i].referencia === "") filas[i].referencia = reg.referencia;
        if (filas[i].tipo === "") filas[i].tipo = reg.tipo;
        valor -= faltaParaCuota;
        saldo = 0;
        i += 1;
        if (i >= filas.length) break;
      }
      if (i >= filas.length) break;

      saldo += valor;
      if (valor > 0) {
        filas[i].valorPagado += valor;
        if (filas[i].tipo === "") filas[i].tipo = reg.tipo;
      }
      if (saldo >= valorCuota) {
        filas[i].fechaPago = reg.fecha;
        if (filas[i].referencia === "") filas[i].referencia = reg.referencia;
        if (filas[i].tipo === "") filas[i].tipo = reg.tipo;
        saldo -= valorCuota;
      } else {
        pagosIdx += 1;
      }
    }
  }

  return filas;
}

/** Equivalente a `calcular_resumen_extracto` en client_report.py */
export function calcularResumenExtracto(
  filas: FilaExtracto[],
  valorCuota: number,
  diasCredito: number = DIAS_CREDITO_DEFAULT,
  totalRegistros?: number,
): Pick<
  MetricasExtracto,
  | "cuotas_generadas"
  | "cuotas_completas"
  | "cuotas_pagadas"
  | "cuotas_pendientes"
  | "total_pagado"
  | "deuda_total"
> {
  const total = filas.reduce((s, f) => s + f.valorPagado, 0);
  let cuotasPagadasCompletas = 0;
  let remanente = 0;
  for (const f of filas) {
    cuotasPagadasCompletas += Math.floor(f.valorPagado / valorCuota);
    remanente += f.valorPagado % valorCuota;
  }
  const fraccionCuota = remanente / valorCuota;
  let cuotasPagadas = cuotasPagadasCompletas + fraccionCuota;
  const cuotasVencidas = Math.min(filas.length, diasCredito);
  if (totalRegistros != null && totalRegistros > 0) {
    cuotasPagadas = Math.max(cuotasPagadas, totalRegistros / valorCuota);
  }
  const cuotasPendientes = Math.max(0, cuotasVencidas - cuotasPagadas);
  const valorPendiente = cuotasPendientes * valorCuota;

  return {
    cuotas_generadas: cuotasVencidas,
    cuotas_completas: cuotasPagadasCompletas,
    cuotas_pagadas: cuotasPagadas,
    cuotas_pendientes: cuotasPendientes,
    total_pagado: total,
    deuda_total: valorPendiente,
  };
}

/** Métricas completas para la web (client_report + campos de mora). */
export function calcularMetricasExtracto(
  fechaInicio: Date,
  valorCuota: number,
  registros: RegistroExtracto[],
  diasCredito: number = DIAS_CREDITO_DEFAULT,
  fechaReferencia?: Date,
  fechasCongeladas: Iterable<string> = [],
): MetricasExtracto {
  const freeze = setCongelados(fechasCongeladas);
  const ref = fechaReferencia ?? new Date();
  const filas = generarFilasExtracto(
    fechaInicio,
    valorCuota,
    registros,
    diasCredito,
    ref,
    freeze,
  );
  const totalRegistros = registros.reduce((s, r) => s + Number(r.valor), 0);
  const resumen = calcularResumenExtracto(
    filas,
    valorCuota,
    diasCredito,
    totalRegistros,
  );

  let ultimoPago = "";
  if (registros.length > 0) {
    const maxFecha = registros.reduce(
      (max, r) => (r.fecha > max ? r.fecha : max),
      registros[0].fecha,
    );
    ultimoPago = formatFecha(startOfDay(maxFecha));
  }

  const fin = startOfDay(ref);
  const diasMora = ultimoPago
    ? diasMoraSinCongelados(ultimoPago, fin, freeze)
    : resumen.cuotas_generadas;

  const cumplimientoPct =
    resumen.cuotas_generadas > 0
      ? Math.round((1000 * resumen.cuotas_pagadas) / resumen.cuotas_generadas) /
        10
      : 0;

  return {
    ...resumen,
    ultimo_pago: ultimoPago,
    dias_mora: diasMora,
    cumplimiento_pct: cumplimientoPct,
  };
}

File: extracto_cliente.py
"""
Métricas del extracto — misma lógica que `mostrar_registros` en `func extrac.txt`.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import date, datetime, timedelta
from typing import Sequence

DIAS_CREDITO_DEFAULT = 365


def parse_dias_credito(fecha_final: str | None = None) -> int:
    raw = (fecha_final or "").strip()
    if raw.isdigit():
        n = int(raw)
        return n if n > 0 else DIAS_CREDITO_DEFAULT
    return DIAS_CREDITO_DEFAULT


@dataclass(frozen=True)
class RegistroExtracto:
    fecha: date
    valor: float
    tipo: str = ""
    referencia: str = ""


@dataclass(frozen=True)
class MetricasExtracto:
    cuotas_generadas: int
    cuotas_completas: int
    cuotas_pagadas: float
    cuotas_pendientes: float
    total_pagado: float
    deuda_total: float
    ultimo_pago: str
    dias_mora: int
    cumplimiento_pct: float


def _as_date(d: date | datetime) -> date:
    if isinstance(d, datetime):
        return d.date()
    return d


def _days_between(a: date, b: date) -> int:
    return (b - a).days


def calcular_metricas_extracto(
    fecha_inicio: date | datetime,
    valor_cuota: float,
    registros: Sequence[RegistroExtracto],
    dias_credito: int = DIAS_CREDITO_DEFAULT,
) -> MetricasExtracto:
    """Replica el reparto de pagos por día del DataFrame en `func extrac.txt`."""
    if valor_cuota <= 0:
        raise ValueError("valor_cuota inválido")

    inicio = _as_date(fecha_inicio)
    fecha_fin_credito = inicio + timedelta(days=dias_credito - 1)
    fin = date.today()
    if fin > fecha_fin_credito:
        fin = fecha_fin_credito

    total = sum(r.valor for r in registros)
    cuotas_pagadas_ceil = math.ceil(total / valor_cuota) if total else 0

    dias_rango = _days_between(inicio, fin) + 1
    if cuotas_pagadas_ceil > dias_rango:
        fin_extendido = fin + timedelta(days=cuotas_pagadas_ceil - dias_rango)
        fin = min(fin_extendido, fecha_fin_credito)

    n = _days_between(inicio, fin) + 1
    valor_pagado = [0.0] * n

    registros_list = [
        RegistroExtracto(
            fecha=_as_date(r.fecha),
            valor=float(r.valor),
            tipo=r.tipo or "",
            referencia=r.referencia or "",
        )
        for r in registros
    ]

    saldo = 0.0
    pagos_idx = 0

    for i in range(n):
        while pagos_idx < len(registros_list) and saldo < valor_cuota:
            valor = registros_list[pagos_idx].valor

            while valor + saldo >= valor_cuota:
                falta_para_cuota = valor_cuota - saldo
                valor_pagado[i] += falta_para_cuota
                valor -= falta_para_cuota
                saldo = 0.0
                i += 1
                if i >= n:
                    break

            if i >= n:
                break

            saldo += valor
            if valor > 0:
                valor_pagado[i] += valor
            if saldo >= valor_cuota:
                saldo -= valor_cuota
            else:
                pagos_idx += 1

    cuotas_pagadas_completas = 0
    remanente = 0.0
    for v in valor_pagado:
        cuotas_pagadas_completas += int(v // valor_cuota)
        remanente += v % valor_cuota

    fraccion_cuota = remanente / valor_cuota
    cuotas_pagadas = cuotas_pagadas_completas + fraccion_cuota
    if total > 0:
        cuotas_pagadas = max(cuotas_pagadas, total / valor_cuota)
    cuotas_vencidas = min(n, dias_credito)
    cuotas_pendientes = max(0.0, cuotas_vencidas - cuotas_pagadas)
    valor_pendiente = cuotas_pendientes * valor_cuota

    ultimo_pago = ""
    if registros_list:
        ultimo_pago = max(r.fecha for r in registros_list).isoformat()

    dias_mora = (
        _days_between(date.fromisoformat(ultimo_pago), fin) if ultimo_pago else cuotas_vencidas
    )

    cumplimiento_pct = (
        round(1000 * cuotas_pagadas / cuotas_vencidas) / 10 if cuotas_vencidas > 0 else 0.0
    )

    return MetricasExtracto(
        cuotas_generadas=cuotas_vencidas,
        cuotas_completas=cuotas_pagadas_completas,
        cuotas_pagadas=cuotas_pagadas,
        cuotas_pendientes=cuotas_pendientes,
        total_pagado=total,
        deuda_total=valor_pendiente,
        ultimo_pago=ultimo_pago,
        dias_mora=dias_mora,
        cumplimiento_pct=cumplimiento_pct,
    )
